### PR TITLE
indexer: use contract `+read` arm to output rice `data`

### PR DESCRIPTION
**Pull request by hosted-fornet**
wip

### CODE REVIEW

**Comment by hosted-fornet** [comment]
```
$ curl --cookie "$ZOD_COOKIE" localhost:8080/~/scry/indexer/json/hash/0xef44.5e1e.2113.c21d.7560.c831.6056.d984.json | jq
{
  "hash": {
    "eggs": {},
    "grains": {
      "0xef44.5e1e.2113.c21d.7560.c831.6056.d984": {
        "location": {
          "epoch-num": 36,
          "block-num": 0,
          "town-id": 1
        },
        "grain": {
          "germ": {
            "salt": 2054606376270393800000,
            "data": {
              "name": "Uqbar DAO",
              "permissions": {
                "host": {
                  "0xef44.5e1e.2113.c21d.7560.c831.6056.d984": [
                    "comms-host"
                  ]
                },
                "read": {
                  "/ship/~zod/dao-links": [
                    "pleb",
                    "owner"
                  ],
                  "/ship/~nec/dao-links": [
                    "pleb",
                    "owner"
                  ],
                  "/ship/~zod/dao-chat": [
                    "pleb",
                    "owner"
                  ],
                  "/ship/~nec/dao-chat": [
                    "pleb",
                    "owner"
                  ]
                },
                "write": {
                  "/ship/~zod/dao-links": [
                    "owner"
                  ],
                  "/ship/~zod/dao-chat": [
                    "pleb",
                    "owner"
                  ],
                  "0xef44.5e1e.2113.c21d.7560.c831.6056.d984": [
                    "owner"
                  ]
                }
              },
              "ship-to-id": {
                "~zod": "0x3.e87b.0cbb.431d.0e8a.2ee2.ac42.d9da.cab8.063d.6bb6.2ff9.b2aa.e1b9.0f56.9c3f.3423",
                "~bus": "0x2.4a1c.4643.b429.dc12.6f3b.03f3.f519.aebb.5439.08d3.e0bf.8fc3.cb52.b92c.9802.636e",
                "~nec": "0x2.eaea.cffd.2bbe.e0c0.02dd.b5f8.dd04.e63f.297f.14cf.d809.b616.2137.126c.da9e.8d3d"
              },
              "subdaos": [],
              "threshold": 0,
              "id-to-ship": {
                "0x3.e87b.0cbb.431d.0e8a.2ee2.ac42.d9da.cab8.063d.6bb6.2ff9.b2aa.e1b9.0f56.9c3f.3423": "~zod",
                "0x2.4a1c.4643.b429.dc12.6f3b.03f3.f519.aebb.5439.08d3.e0bf.8fc3.cb52.b92c.9802.636e": "~bus",
                "0x2.eaea.cffd.2bbe.e0c0.02dd.b5f8.dd04.e63f.297f.14cf.d809.b616.2137.126c.da9e.8d3d": "~nec"
              },
              "members": {
                "0x3.e87b.0cbb.431d.0e8a.2ee2.ac42.d9da.cab8.063d.6bb6.2ff9.b2aa.e1b9.0f56.9c3f.3423": [
                  "comms-host",
                  "owner"
                ],
                "0x2.4a1c.4643.b429.dc12.6f3b.03f3.f519.aebb.5439.08d3.e0bf.8fc3.cb52.b92c.9802.636e": [
                  "pleb"
                ],
                "0x2.eaea.cffd.2bbe.e0c0.02dd.b5f8.dd04.e63f.297f.14cf.d809.b616.2137.126c.da9e.8d3d": [
                  "pleb"
                ]
              },
              "proposals": {}
            },
            "is-rice": true
          },
          "lord": "0x6f.6164",
          "id": "0xef44.5e1e.2113.c21d.7560.c831.6056.d984",
          "holder": "0x6f.6164",
          "town-id": 1
        }
      }
    },
    "slots": {}
  }
}
```

**Comment by hosted-fornet** [comment]
6d4c518 adds zigs balances and metadata:

```json
$ curl --cookie "$ZOD_COOKIE" localhost:8080/~/scry/indexer/json/hash/0xcb22.90ff.6e2f.ba63.86e1.40bb.6e89.c2d8.json | jq
{
  "hash": {
    "eggs": {},
    "grains": {
      "0xcb22.90ff.6e2f.ba63.86e1.40bb.6e89.c2d8": {
        "location": {
          "epoch-num": 5,
          "block-num": 0,
          "town-id": 1
        },
        "grain": {
          "germ": {
            "salt": 1936157050,
            "data": {
              "balance": 10321055000000000000,
              "metadata": "0x61.7461.6461.7465.6d2d.7367.697a",
              "allowances": {}
            },
            "is-rice": true
          },
          "lord": "0x74.6361.7274.6e6f.632d.7367.697a",
          "id": "0xcb22.90ff.6e2f.ba63.86e1.40bb.6e89.c2d8",
          "holder": "0x3.e87b.0cbb.431d.0e8a.2ee2.ac42.d9da.cab8.063d.6bb6.2ff9.b2aa.e1b9.0f56.9c3f.3423",
          "town-id": 1
        }
      }
    },
    "slots": {}
  }
}
```

```json
$ curl --cookie "$ZOD_COOKIE" localhost:8080/~/scry/indexer/json/hash/0x61.7461.6461.7465.6d2d.7367.697a.json | jq
{
  "hash": {
    "eggs": {},
    "grains": {
      "0x61.7461.6461.7465.6d2d.7367.697a": {
        "location": {
          "epoch-num": 5,
          "block-num": 0,
          "town-id": 0
        },
        "grain": {
          "germ": {
            "salt": 1936157050,
            "data": {
              "name": "Uqbar Tokens",
              "salt": 1936157050,
              "decimals": 18,
              "deployer": "0x0",
              "cap": null,
              "minters": [],
              "symbol": "ZIG",
              "supply": 1e+24,
              "mintable": false
            },
            "is-rice": true
          },
          "lord": "0x74.6361.7274.6e6f.632d.7367.697a",
          "id": "0x61.7461.6461.7465.6d2d.7367.697a",
          "holder": "0x74.6361.7274.6e6f.632d.7367.697a",
          "town-id": 0
        }
      }
    },
    "slots": {}
  }
}
```

**Comment by hosted-fornet** [comment]
41295a8 enables eggs args (i.e. tx args) scries in general, and specifically implements it for zigs. Example:

```json
$ curl --cookie "$ZOD_COOKIE" localhost:8080/~/scry/indexer/json/hash/0x3.e87b.0cbb.431d.0e8a.2ee2.ac42.d9da.cab8.063d.6bb6.2ff9.b2aa.e1b9.0f56.9c3f.3423.json | jq
{
  "hash": {
    "eggs": {
      ...,
      "0x2f62.ec20.3687.5763.ffd0.2d44.300f.79b7.1e08.8aee.702f.80af.b748.c7f7.21d0.4e24": {
        "location": {
          "epoch-num": 5,
          "egg-num": 0,
          "block-num": 0,
          "town-id": 0
        },
        "egg": {
          "shell": {
            "status": 0,
            "budget": 10000,
            "from": {
              "id": "0x3.e87b.0cbb.431d.0e8a.2ee2.ac42.d9da.cab8.063d.6bb6.2ff9.b2aa.e1b9.0f56.9c3f.3423",
              "nonce": 2,
              "zigs": 1.10558721324506e+38
            },
            "to": "0x74.6361.7274.6e6f.632d.7367.697a",
            "eth-hash": null,
            "rate": 1,
            "town-id": 0,
            "sig": {
              "life": 1.2850776036546031e+76,
              "ship": "~nidsyt-parler-foldet-rannet--dapfed-digdec-lasfer-loplep--tappub-nisfed-sopsyt-nopwyd--rivnet-mordeg-litnux-posnet",
              "hash": "0x1"
            }
          },
          "yolk": {
            "args": {
              "give": {
                "budget": 10000,
                "to": "0x2.eaea.cffd.2bbe.e0c0.02dd.b5f8.dd04.e63f.297f.14cf.d809.b616.2137.126c.da9e.8d3d",
                "amount": 777,
                "account": null
              }
            },
            "my-grains": [
              "0x532c.d5cf.befc.5c0f.86e1.40bb.6e89.c2d8"
            ],
            "caller": {
              "id": "0x3.e87b.0cbb.431d.0e8a.2ee2.ac42.d9da.cab8.063d.6bb6.2ff9.b2aa.e1b9.0f56.9c3f.3423",
              "nonce": 2,
              "zigs": 1.10558721324506e+38
            },
            "cont-grains": []
          }
        }
      }
    },
    "grains": {},
    "slots": {}
  }
}
```

**Comment by hosted-fornet** [comment]
Added `+read` arms to fungible.hoon and nft.hoon in ee45191 and f489296. Have not yet tested.

**Comment by hosted-fornet** [comment]
Current status:
1. zigs: rice data and egg args enabled
2. fungible: rice data and egg args enabled
3. nft: rice data and egg args enabled
4. capitol: code written but not working

**Comment by hosted-fornet** [comment]
97044ca fixes bug that was stopping the capitol contract rice data & egg args from being scried out (had been scrying only %sequencer).

**Comment by hosted-fornet** [comment]
I recently upgraded my fakezods to 1.9 and now the `/json/hash` scry is not working: http spinning forever and spamming

```
fund: in dec, parent 6a4f0dca not found at 7
```

Not sure if axes, or something like that, got changed in the upgrade?

Anyways, aside from the issue about testing, the only other thing I want to do before review is to update the DAO tests so the read test will pass.

**Comment by hosted-fornet** [comment]
@tacryt-socryp going to retarget this PR to PR #52 and do some refactoring. I'll ping you again here when ready for review post-refactor.

**Comment by hosted-fornet** [comment]
`+read` arm calls working again post-PR #52. Ready for review @tacryt-socryp

File: lib/zig/contracts/fungible.hoon [hunk 1]
```diff
@@ -279,9 +279,197 @@
...
+      ::
+      ++  metadata  ::  TODO: grab token-metadata?
+        |=  md-id=id
+        [%s (scot %ux md-id)]
+      ::
+      --
+    ::
+    ++  enjs-token-metadata
+      =,  enjs:format
+      |^
```
Comment (tacryt-socryp) [inline lib/zig/contracts/fungible.hoon:327]: No need for `|^` here

File: lib/zig/contracts/fungible.hoon [hunk 2]
```diff
@@ -279,9 +279,197 @@
...
+        |=  set-id-bal=(set [id @ud])
+        ^-  ^json
+        :-  %a
+        %+  turn  ~(tap in set-id-bal)
+        |=  [i=id bal=@ud]
+        %-  pairs
+        :+  [%id %s (scot %ux i)]
+          [%bal (numb bal)]
+        ~
+      ::
```
Comment (tacryt-socryp) [inline lib/zig/contracts/fungible.hoon:414]: Note that `::` blocks at the beginning and end of cores are considered unnecessary.

File: lib/zig/util.hoon [hunk 3]
```diff
@@ -246,15 +252,32 @@
...
     ?.  ?=(%& -.germ.u.found)           ~
     ?.  =(lord.u.found id)              ~
-    `[find u.res]
+    `[find u.found]
   ::  this isn't an ideal method but okay for now
   ::  goal is to return ~ if some rice weren't found
   ?.  =(~(wyt by owns) (lent contract-rice))
     ``noun+!>(~)
-  =/  cont  !<(contract:smart [-:!>(*contract:smart) u.cont.p.germ.u.res])  
+  =/  library  ;;(vase (cue q.q.smart-lib))
```
Comment (tacryt-socryp) [inline lib/zig/util.hoon:260]: Important to avoid doing this every time. Can you either pass this in via a door or might be possible with `^~`

**Review by tacryt-socryp — CHANGES_REQUESTED** [review]
Overall looks good, a few spots stood out as unperformant.

File: lib/zig/contracts/fungible.hoon (hunk 1 above)
Comment (hosted-fornet) [reply lib/zig/contracts/fungible.hoon]: Removed in 9a1e4fc

File: lib/zig/contracts/fungible.hoon (hunk 2 above)
Comment (hosted-fornet) [reply lib/zig/contracts/fungible.hoon]: Removed from contracts in 9a1e4fc

File: lib/zig/contracts/capitol.hoon [hunk 4]
```diff
@@ -94,9 +94,116 @@
...
   ++  json
-    ~
+    |^  ^-  ^json
+    ?+    args  !!
+    ::
+        [%rice-data ~]
+      ?>  =(1 ~(wyt by owns.cart))
+      =/  g=grain  -:~(val by owns.cart)
+      ?>  ?=(%& -.germ.g)
+      =+  [is-ziggurat *]=(mule |.(;;(ziggurat data.p.germ.g)))
```
[the remark this answers is not in the document]
Comment (hosted-fornet) [reply lib/zig/contracts/capitol.hoon]: Improved type branching (no use of `mule`) in 66a9883

File: lib/zig/util.hoon (hunk 3 above)
Comment (hosted-fornet) [reply lib/zig/util.hoon]: See 8087fc1 where I pass in via a door.

NOTE: I change what we store in `app/ziggurat.hoon` and `app/sequencer.hoon`: this assumes the expensive part is the `;;` and not plugging the `vase` into the door. Please confirm @tacryt-socryp

**Review by tacryt-socryp — APPROVED** [review]
Perfect LGTM

File: lib/zig/util.hoon (hunk 3 above)
Comment (tacryt-socryp) [reply lib/zig/util.hoon]: ;; expensive

**Pull request merged**